timeutil/timer.go

```diff
@@ -0,0 +1,178 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package timer provides a wrapper around time.Timer and its associated
+// functions so that timers can be mocked in tests.
+package timeutil
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+// Timer is an interface which wraps time.Timer so that it may be mocked.
+type Timer interface {
+	Reset(d time.Duration) bool
+	Stop() bool
+}
+
+// FakeAfterFunc creates a new fake timer which will call the given callback in
+// its own goroutine when the timer fires. The returned FakeTimer's C field is
+// not used and will be nil.
+//
+// This simulates the behavior of AfterFunc() from the time package.
+// See here for more details: https://pkg.go.dev/time#AfterFunc
+func FakeAfterFunc(d time.Duration, f func()) *FakeTimer {
+	currTime := time.Now()
+	return &FakeTimer{
+		currTime:   currTime,
+		expiration: currTime.Add(d),
+		callback:   f,
+	}
+}
+
+// FakeNewTimer creates a new fake timer which, when it fires, will send the
+// time that the timer fires over the C channel.
+//
+// This simulates the behavior of NewTimer() from the time package.
+// See here for more details: https://pkg.go.dev/time#NewTimer
+func FakeNewTimer(d time.Duration) *FakeTimer {
```

Comment:
did you mean NewFakeTimer?

Reply:
I would suggest to name the package testtime and name all types and functions just as they show up in the time package. Specifically I would avoid the words test and fake, outside of the package name.

Reply:
What about the interface itself? I think we should make it clear that the `Timer` interface should be used with `time.Timer` in production and something like a `testtime.TestTimer` in tests, and that people shouldn't be using `testtime.NewTimer` or `testtime.AfterFunc` in non-test code. I'm not quite sure how best to do that... We could have `NewTimer` and `AfterFunc` call `time.NewTimer` and `time.AfterFunc` unless they're mocked, but then we'd need global variables akin to how logger is mocked, and I'd rather avoid something unusual like that.

Perhaps there's some way to cause compilation to fail if one tries to use `testtime.NewTimer` or `testtime.AfterFunc` outside of test code? Or we could add it as a static check?

Reply:
there's osutil.MustBeTestBinary()

Reply:
I added a static check (before I saw this), but can use this instead if preferred.